File: 2016/01/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};
use regex::Regex;
use num::complex::Complex;

#[derive(Debug)]
struct Instruction {
    direction: char,
    distance: u32,
}

fn get_manhatan_distance(position: &Complex<i32>) -> i32 {
    i32::abs(position.re) + i32::abs(position.im)
}

fn get_new_heading(current_heading: Complex<i32>, direction: char) -> Complex<i32> {
    match direction {
        'L' => current_heading * Complex::new(0, 1),
        'R' => current_heading * Complex::new(0, -1),
        _ => panic!("Unknow direction")
    }
}

fn solve(instructions: &Vec<Instruction>) -> (i32,i32) {
    let mut position = Complex::new(0, 0);
    let mut heading = Complex::new(0, 1);
    let mut first_repeated = 0;
    let mut visited_positions = Vec::new();
    for instruction in instructions {
        heading = get_new_heading(heading, instruction.direction);
        for _ in 0..instruction.distance {
            position += heading;
            if first_repeated == 0 {
                if visited_positions.contains(&position) {
                    first_repeated = get_manhatan_distance(&position);
                } else {
                    visited_positions.push(position);
                }
            }
        }
    }
    (get_manhatan_distance(&position), first_repeated)
}

fn get_input(file_path: &String) -> Vec<Instruction> {
    let re = Regex::new(r"(?P<direction>[RL])(?P<distance>\d+),?\s?").unwrap();
    fs::read_to_string(file_path).expect("Error reading input file").split(" ")
        .map(|split| re.captures(split)
            .and_then(|cap| Some(
                    (
                        cap.name("direction").map(|c| c.as_str().chars().next().unwrap()).unwrap(), 
                        cap.name("distance").map(|c| c.as_str()).unwrap()
                    )
                )
            )
        ).map(|caps|
            match caps {
                Some((direction, distance)) => Instruction{
                    direction, 
                    distance: distance.parse().expect("Unable to parse distance")
                },
                _ => panic!("Unable to parse instructions")
            }
        ).collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:?}", now.elapsed().as_secs_f32());
}

File: 2016/01/cpp/main.cpp
#include <iostream>
#include <fstream>
#include <ctime>
#include <regex>
#include <vector>
#include <complex>
#include <cmath>

using namespace std;
using namespace std::complex_literals;

struct Results
{
    int part1;
    int part2;
};

struct Instruction
{
    char direction;
    int distance;
    Instruction(char dir, int dis)
    {
        direction = dir;
        distance = dis;
    }
};

int getManhatanDistance(complex<double> position)
{
    return abs<double>(position.real()) + abs<double>(position.imag());
}

Results solve(vector<Instruction> instructions)
{
    complex<double> position;
    complex<double> heading = 1i;
    int part2 = 0;
    vector<complex<double>> visited;
    for (Instruction instruction : instructions)
    {
        heading *= instruction.direction == 'L' ? 1i : -1i;
        for (auto i = 0; i < instruction.distance; i++)
        {
            position += heading;
            if (part2 == 0)
            {
                if (find(visited.begin(), visited.end(), position) != visited.end())
                {
                    part2 = getManhatanDistance(position);
                }
                else
                {
                    visited.push_back(position);
                }
            }
        }
    }
    return {getManhatanDistance(position), part2};
}

vector<Instruction> getInput(char *filePath)
{
    ifstream file(filePath);
    if (!file.is_open())
        throw runtime_error("Error reading input file!");

    string content((istreambuf_iterator<char>(file)), (istreambuf_iterator<char>()));
    file.close();
    regex instructions_regex("([RL])(\\d+),?\\s?");
    vector<Instruction> instructions;
    for (sregex_iterator i = sregex_iterator(content.begin(), content.end(), instructions_regex);
         i != sregex_iterator(); i++)
    {
        smatch match = *i;
        instructions.push_back(Instruction(match.str(1)[0], stoi(match.str(2))));
    }
    return instructions;
}

int main(int argc, char *argv[])
{
    if (argc != 2)
        throw runtime_error("Please, add input file path as parameter");

    clock_t begin = clock();
    auto results = solve(getInput(argv[1]));
    clock_t end = clock();
    auto elapsed_secs = double(end - begin) / CLOCKS_PER_SEC;
    cout << "P1: " << results.part1 << endl;
    cout << "P2: " << results.part2 << endl;
    cout << endl;
    cout.precision(7);
    cout << "Time: " << std::fixed << elapsed_secs << endl;
    return 0;
}